### PR TITLE
New Module Hash Model

### PR DESCRIPTION
**Pull request by rachel-bousfield**
This PR leverages the new onchain module hash to make a number of simplifications and improvements
- We no longer tie activation to account objects. You'll notice in the Geth PR it's entirely module hash based.
- Validators now store activated prover Module's in the db, which they pre-fetch before fraud proofs & block validation.
- JIT validation no longer requires any pre-activation. We load existing Asm from disk.
- Activation pricing now happens in Rust
- We patch ToB's table size concern.
- Other efficiency improvements, for example reducing the number of Asm copies.

This PR includes conceptual changes that will be easier to track by adopting the following nomenclature.
- **activation** the process of instrumenting a WASM to produce an **asm** and **module**
- **asm** the instrumented, x86 or ARM version of a Stylus contract nodes run in the happy case
- **module** the instrumented, WAVM version of a Stylus contract nodes link-in when fraud proving
- **module hash** the hash of a **module**, which uniquely determines a **module** and **asm** pair

Associated PRs
- https://github.com/OffchainLabs/stylus-geth/pull/29
- https://github.com/OffchainLabs/stylus/pull/95

### CODE REVIEW

File: arbitrator/wasm-libraries/user-host/src/link.rs [hunk 1]
```diff
@@ -35,60 +35,47 @@ extern "C" {
 #[repr(C, align(256))]
 struct MemoryLeaf([u8; 32]);
 
-/// Compiles and instruments a user wasm.
+/// Instruments a user wasm.
```
Comment (joshuacolvin0) [inline arbitrator/wasm-libraries/user-host/src/link.rs]: `Instruments and activates`?

File: arbitrator/wasm-libraries/user-host/src/link.rs [hunk 2]
```diff
@@ -35,60 +35,47 @@ extern "C" {
...
     sp: usize,
 ) {
     let mut sp = GoStack::new(sp);
     let wasm = sp.read_go_slice_owned();
     let page_limit = sp.read_u16();
     let version = sp.read_u16();
     let debug = sp.read_bool32();
-    let (out_hash_ptr, out_hash_len) = sp.read_go_slice();
+    let module_hash = sp.read_go_ptr();
+    let gas = sp.read_go_ptr();
```
Comment (joshuacolvin0) [inline arbitrator/wasm-libraries/user-host/src/link.rs:63]: maybe add comment about `modules_hash` being output, `gas` being input & output

File: arbos/programs/programs.go [hunk 3]
```diff
@@ -342,7 +337,7 @@ func (p Programs) setProgram(codehash common.Hash, program Program) error {
 	if err != nil {
 		return err
 	}
-	return p.compiledHashes.Set(codehash, program.compiledHash)
+	return p.compiledHashes.Set(codehash, program.moduleHash)
```
Comment (joshuacolvin0) [inline arbos/programs/programs.go]: update `p.compiledHashes` to `p.moduleHashes`?

**Review by joshuacolvin0 — APPROVED** [review]
LGTM

**Review by joshuacolvin0 — APPROVED** [review]
LGTM

**Review by joshuacolvin0 — APPROVED** [review]
LGTM

**Review by joshuacolvin0 — APPROVED** [review]
LGTM

File: arbos/programs/native.go [hunk 4]
```diff
@@ -93,6 +93,7 @@ func activateProgram(
 func callProgram(
 	address common.Address,
 	program Program,
```
Comment (PlasmaPower) [inline arbos/programs/native.go]: This `program` parameter is no longer used now that `moduleHash` is passed in separately.

File: arbos/burn/burn.go [hunk 5]
```diff
@@ -13,7 +13,7 @@ import (
 type Burner interface {
 	Burn(amount uint64) error
 	Burned() uint64
-	GasLeft() *uint64
+	GasLeft() *uint64 // SystemBurner's panic
```
Comment (PlasmaPower) [inline arbos/burn/burn.go]: The apostrophe doesn't read well to me, could you do
```
// `SystemBurner`s panic
```
instead?

**Review by PlasmaPower — COMMENTED** [review]
A couple minor additional comments but largely LGTM

**Review by PlasmaPower — APPROVED** [review]
LGTM

**Pull request merged**